# Recursion

Factorial

In [1]:
def fact(n):
    if n == 0:
        return 0
    elif n == 1 :
        return 1
    else :
        return n * (fact(n-1))


fact(5)

120

Is Palindrome

In [2]:
def is_palindrome(s):
    if len(s) <= 1:
        return True
    if s[0] != s[-1]:
        return False
    return is_palindrome(s[1:-1])

is_palindrome("ababsg")

False

In [3]:
def bitStrings(n): 
    if n == 0: return [] 
    if n == 1: return ["0", "1"]
    return [
        bitstring + digit for digit in bitStrings(1)
        for bitstring in bitStrings(n-1)
    ]
    
print(bitStrings(3))

['000', '100', '010', '110', '001', '101', '011', '111']


# Linked List

In [4]:
class Node:
    def __init__(self, data = None):
        self.data = data
        self.next = None
        self.prev = None

- **Singly linked list**

In [5]:
from collections.abc import Iterable

class LinkedList:
    def __init__(self, values = None):
        self.head = None
        if values:
            for val in values:
                self.append(val)

# Append a single element to array  
    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

# Append an iterable
    def extend(self, iterable):
        if not isinstance(iterable, Iterable):
            raise TypeError("extend() method only works with an iterable")
        for val in iterable:
            self.append(val)
            
#Join two linked lists
    def join(self, ll):
        if not isinstance(ll, LinkedList):
            raise ValueError("join() expects another linked list")
        if not self.head:
            self.head = ll.head
            return
        if not ll.head:
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = ll.head
    
# Insert an element in between
    def insert(self, index, value):
        new_node = Node(value)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return

        count = 0
        current = self.head

        while count < index - 1 :
            if not current:
                raise IndexError("Linked list index out of bounds")
            current = current.next
            count +=1 


        new_node.next = current.next
        current.next = new_node

# Insert an iterable
    def insert_many(self, index, iterable):
        if not isinstance(iterable, Iterable):
            raise ValueError("insert_many() expects an iterable value")

        nodes = [Node(val) for val in iterable]
        for i in range(len(nodes) - 1):
            nodes[i].next = nodes[i+1]

        if index == 0:
            nodes[-1].next = self.head
            self.head = nodes[0]
            return

        count = 0
        current = self.head

        while count < index - 1:
            if not current:
                raise IndexError("Linked List index out of bounds")
            current = current.next
            count += 1

        nodes[-1].next = current.next
        current.next = nodes[0]

# Delete an element by index
    def delete(self, index):
        if not self.head:
            raise IndexError("Can't delete from empty list")
        
        if index == 0:
            self.head = self.head.next
            return

        current = self.head
        count = 0
        while count < index - 1:
            if not current:
                raise IndexError("Linked list index out of bounds")
            current = current.next
            count += 1 
                
        current.next = current.next.next 

#Delete first occurence of an element by value
    def remove(self, value):
        current = self.head
        previous = None
        while current:
            if current.data == value:
                if not previous:
                    self.head = current.next
                else:
                    previous.next = current.next
                return

            previous = current
            current = current.next
        raise ValueError(f"{value} not found in the Linked List")

# Sort the linked list
    def sort(self, reverse = False):
        values = list(self)
        values.sort(reverse = reverse)
        self.head = None
        for val in values:
            self.append(val)

# Reverse the linked list
    def reverse(self):
        prev = None
        current = self.head
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        self.head = prev   

# Remove duplicates
    def remove_dups(self):
        seen = set()
        current = self.head
        prev = None
        while current:
            if current.data in seen:
                prev.next = current.next
            else:
                seen.add(current.data)

            prev = current
            current = current.next
            

    
    def __getitem__(self, index):
        count = 0
        current = self.head
        while current:
            if count == index:
                return current.data
            count += 1
            current = current.next
        raise IndexError("Linked list index out of bounds")

    def __setitem__(self, index, value):
        count = 0
        current = self.head
        while current:
            if count == index:
                current.data = value
                return
            current = current.next
            count += 1
        raise IndexError("Linked list index out of bounds")

    def __len__(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count

    def __iter__(self):
        current = self.head
        while current:
            yield current.data
            current = current.next

    def __repr__(self):
        elements = []
        current = self.head
        while current:
            elements.append(str(current.data))
            current = current.next
        return "[" + " -> ".join(elements) + "]"




In [6]:
def test_linked_list():
    print("----- Testing LinkedList -----")

    # Create linked list with initial values
    ll = LinkedList([10, 20, 30])
    print("Initial list:")
    print(ll)

    # Append single element
    ll.append(40)
    print("\nAfter append(40):")
    print(ll)

    # Extend with multiple elements
    ll.extend([50, 60])
    print("\nAfter extend([50, 60]):")
    print(ll)

    # Insert element at index
    ll.insert(2, 25)
    print("\nAfter insert(2, 25):")
    print(ll)

    # Insert multiple elements at index
    ll.insert_many(4, [26, 27, 28])
    print("\nAfter insert_many(4, [26,27,28]):")
    print(ll)

    # Delete element by index
    ll.delete(3)
    print("\nAfter delete(3):")
    print(ll)

    # Remove element by value
    ll.remove(27)
    print("\nAfter remove(27):")
    print(ll)

    # Get item
    print("\nElement at index 2:", ll[2])

    # Set item
    ll[2] = 99
    print("After setting ll[2] = 99:")
    print(ll)

    # Length
    print("\nLength of list:", len(ll))

    # Sort ascending
    ll.sort()
    print("\nAfter sort():")
    print(ll)

    # Sort descending
    ll.sort(reverse=True)
    print("\nAfter sort(reverse=True):")
    print(ll)

    # Reverse list
    ll.reverse()
    print("\nAfter reverse():")
    print(ll)

    # Iterating using for loop
    print("\nIterating using for loop:")
    for val in ll:
        print(val, end=" -> ")
    print("None")

    print("\n----- Test Completed -----")


test_linked_list()

----- Testing LinkedList -----
Initial list:
[10 -> 20 -> 30]

After append(40):
[10 -> 20 -> 30 -> 40]

After extend([50, 60]):
[10 -> 20 -> 30 -> 40 -> 50 -> 60]

After insert(2, 25):
[10 -> 20 -> 25 -> 30 -> 40 -> 50 -> 60]

After insert_many(4, [26,27,28]):
[10 -> 20 -> 25 -> 30 -> 26 -> 27 -> 28 -> 40 -> 50 -> 60]

After delete(3):
[10 -> 20 -> 25 -> 26 -> 27 -> 28 -> 40 -> 50 -> 60]

After remove(27):
[10 -> 20 -> 25 -> 26 -> 28 -> 40 -> 50 -> 60]

Element at index 2: 25
After setting ll[2] = 99:
[10 -> 20 -> 99 -> 26 -> 28 -> 40 -> 50 -> 60]

Length of list: 8

After sort():
[10 -> 20 -> 26 -> 28 -> 40 -> 50 -> 60 -> 99]

After sort(reverse=True):
[99 -> 60 -> 50 -> 40 -> 28 -> 26 -> 20 -> 10]

After reverse():
[10 -> 20 -> 26 -> 28 -> 40 -> 50 -> 60 -> 99]

Iterating using for loop:
10 -> 20 -> 26 -> 28 -> 40 -> 50 -> 60 -> 99 -> None

----- Test Completed -----


- **Doubly linked list**

In [7]:
from collections.abc import Iterable

class DoublyLinkedList:
    def __init__(self, values = None):
        self.head = None
        self.tail = None
        if values:
            for val in values:
                self.append(val)

# Append a single value
    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = self.tail = new_node
            return
        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node

# Append an iterable
    def extend(self, iterable):
        if not isinstance(iterable, Iterable):
            raise ValueError("extend() method only works with an iterable")
        for val in iterable:
            self.append(val)

# Insert an element at given index
    def insert(self, index, data):
        new_node = Node(data)
        
        if index == 0:
            new_node.next = self.head
            if self.head:
                self.head.prev = new_node
            else:
                self.tail = new_node
            self.head = new_node
            return

        current = self.head
        count = 0

        while count < index - 1:
            if not current:
                raise IndexError("Linked list index out of bounds")
            current = current.next
            count += 1

        new_node.next = current.next
        new_node.prev = current

        if current.next:
            current.next.prev = new_node
        else:
            self.tail = new_node
        current.next = new_node

# Insert an iterable at given index
    def insert_many(self, index, iterable):
        if not isinstance(iterable, Iterable):
            raise ValueError("insert_many() expects an iterable input")

        nodes = [Node(val) for val in iterable]
        for i in range(len(nodes)-1):
            nodes[i].next, nodes[i+1].prev = nodes[i+1], nodes[i]

        if index == 0:
            nodes[-1].next = self.head
            if self.head:
                self.head.prev = nodes[-1]
            else:
                self.tail = nodes[-1]
            self.head = nodes[0]
            return

        current = self.head
        count = 0
        
        while count < index-1:
            if not current:
                raise IndexError("Index of the linked list out of bounds")
            current = current.next
            count += 1

        nodes[-1].next = current.next
        nodes[0].prev = current

        if current.next:
            current.next.prev = nodes[-1]
        else:
            self.tail = nodes[-1]
        current.next = nodes[0]
        
# Delete an item by index
    def delete(self, index):
        if not self.head:
            raise IndexError("Cannot delete from empty list")
        
        current = self.head
        count = 0
        while count < index :
            if not current:
                raise IndexError("Linked List index out of bounds")
            current = current.next
            count += 1

        if current.prev:
            current.prev.next = current.next
        else:
            self.head = current.next

        if current.next:
            current.next.prev = current.prev
        else:
            self.tail = current.prev
        
# Delete the first occurence of an element
    def remove(self, value):
        if not self.head:
            raise IndexError("Cannot delete from empty list")

        current = self.head
        while current:
            if current.data == value:
                if current.prev:
                    current.prev.next = current.next
                else:
                    self.head = current.next

                if current.next:
                    current.next.prev = current.prev
                else :
                    self.tail = current.prev
                return
            current = current.next
        raise ValueError(f"{value} not found in the list")


# Sort the linked list
    def sort(self, reverse=False):
        values =[]
        current = self.head
        while current:
            values.append(current.data)
            current = current.next
        
        values.sort(reverse = reverse)
        self.head = self.tail = None
        for val in values:
            self.append(val)


# Reverse the linked list
    def reverse(self):
        current = self.head
        while current:
            current.next, current.prev = current.prev, current.next
            if not current.prev:
                self.tail = self.head
                self.head = current
            current = current.prev

    def __getitem__(self, index):
        current = self.head
        count = 0
        while current:
            if count == index:
                return current.data
            current = current.next
            count += 1
        raise IndexError("Linked List Index out of bounds")


    def __setitem__(self, index, value):
        current = self.head
        count = 0
        while current:
            if count == index:
                current.data = value
                return
            current = current.next
            count += 1
        raise IndexError("Index out of bounds")


    def __len__(self):
        count = 0 
        current = self.head
        while current:
            current = current.next
            count += 1
        return count


    def __iter__(self):
        current = self.head
        while current:
            yield current.data
            current = current.next

        
    def __repr__(self):
        elements = []
        current = self.head
        while current:
            elements.append(str(current.data))
            current = current.next

        return "[" + " <-> ".join(elements) + "]"

In [8]:
def test_doubly_linked_list():
    print("=== Creating list with initial values ===")
    dll = DoublyLinkedList([1, 2, 3])
    print(dll)  # [1 <-> 2 <-> 3]

    print("\n=== Append a single value ===")
    dll.append(4)
    print(dll)  # [1 <-> 2 <-> 3 <-> 4]

    print("\n=== Extend with an iterable ===")
    dll.extend([5, 6])
    print(dll)  # [1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6]

    print("\n=== Insert at index ===")
    dll.insert(0, 0)
    dll.insert(3, 99)
    print(dll)  # [0 <-> 1 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

    print("\n=== Insert many at index ===")
    dll.insert_many(2, [7, 8])
    print(dll)  # [0 <-> 1 <-> 7 <-> 8 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

    print("\n=== Delete by index ===")
    dll.delete(3)  # remove 8
    print(dll)  # [0 <-> 1 <-> 7 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

    print("\n=== Remove by value ===")
    dll.remove(99)
    print(dll)  # [0 <-> 1 <-> 7 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6]

    print("\n=== Reverse the list ===")
    dll.reverse()
    print(dll)  # [6 <-> 5 <-> 4 <-> 3 <-> 2 <-> 7 <-> 1 <-> 0]

    print("\n=== Sort the list ===")
    dll.sort()
    print(dll)  # [0 <-> 1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6 <-> 7]

    print("\n=== Indexing ===")
    print(dll[0])  # 0
    print(dll[7])  # 7

    print("\n=== Setting values by index ===")
    dll[0] = 100
    dll[7] = 200
    print(dll)  # [100 <-> 1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6 <-> 200]

    print("\n=== Iterating over the list ===")
    for val in dll:
        print(val, end=" ")  # 100 1 2 3 4 5 6 200
    print()

    print("\n=== Length of the list ===")
    print(len(dll))  # 8
test_doubly_linked_list()

=== Creating list with initial values ===
[1 <-> 2 <-> 3]

=== Append a single value ===
[1 <-> 2 <-> 3 <-> 4]

=== Extend with an iterable ===
[1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6]

=== Insert at index ===
[0 <-> 1 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

=== Insert many at index ===
[0 <-> 1 <-> 7 <-> 8 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

=== Delete by index ===
[0 <-> 1 <-> 7 <-> 2 <-> 99 <-> 3 <-> 4 <-> 5 <-> 6]

=== Remove by value ===
[0 <-> 1 <-> 7 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6]

=== Reverse the list ===
[6 <-> 5 <-> 4 <-> 3 <-> 2 <-> 7 <-> 1 <-> 0]

=== Sort the list ===
[0 <-> 1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6 <-> 7]

=== Indexing ===
0
7

=== Setting values by index ===
[100 <-> 1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 6 <-> 200]

=== Iterating over the list ===
100 1 2 3 4 5 6 200 

=== Length of the list ===
8


In [9]:
def test_doubly_linked_list_with_ack():
    # Helper function to convert DLL to Python list
    def to_list(dll):
        return list(dll)
    
    dll = DoublyLinkedList([1, 2, 3])
    assert to_list(dll) == [1, 2, 3]

    dll.append(4)
    assert to_list(dll) == [1, 2, 3, 4]

    dll.extend([5, 6])
    assert to_list(dll) == [1, 2, 3, 4, 5, 6]

    dll.insert(0, 0)
    dll.insert(3, 99)
    assert to_list(dll) == [0, 1, 2, 99, 3, 4, 5, 6]

    dll.insert_many(2, [7, 8])
    assert to_list(dll) == [0, 1, 7, 8, 2, 99, 3, 4, 5, 6]

    dll.delete(3)  # remove 8
    assert to_list(dll) == [0, 1, 7, 2, 99, 3, 4, 5, 6]

    dll.remove(99)
    assert to_list(dll) == [0, 1, 7, 2, 3, 4, 5, 6]

    dll.reverse()
    assert to_list(dll) == [6, 5, 4, 3, 2, 7, 1, 0]

    dll.sort()
    assert to_list(dll) == [0, 1, 2, 3, 4, 5, 6, 7]

    assert dll[0] == 0
    assert dll[7] == 7

    dll[0] = 100
    dll[7] = 200
    assert to_list(dll) == [100, 1, 2, 3, 4, 5, 6, 200]

    # Iteration test
    iterated = [x for x in dll]
    assert iterated == [100, 1, 2, 3, 4, 5, 6, 200]

    assert len(dll) == 8

    # If all asserts pass
    print("✅ All test cases passed successfully!")


test_doubly_linked_list_with_ack()


✅ All test cases passed successfully!


- **Circular Linked List**

In [10]:
from collections.abc import Iterable

class CircularLinkedList:
    def __init__(self, values=None):
        self.head = None
        if values:
            for val in values:
                self.append(val)

# Append a single node
    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            new_node.next =  self.head
            return

        current = self.head
        while current.next != self.head:
            current = current.next

        current.next = new_node
        new_node.next = self.head

# Append multiple elements (iterable)
    def extend(self, iterable):
        if not isinstance(iterable, Iterable):
            raise ValueError("extend() expects an iterable as input")
        for val in iterable:
            self.append(val)

# Insert at a specific index
    def insert(self, index, value):
        new_node = Node(value)

        current = self.head
        count = 0
        while count < index-1:
            if current.next == self.head:
                raise IndexError("Index out of bounds")
            current = current.next
            count += 1

        current.next, new_node.next = new_node, current.next

    def __iter__(self):
        current = self.head
        while True:
            yield current.data
            current = current.next
            if current == self.head:
                break

            



Task: Find the Smallest Missing Positive Number in a Sorted Array

Problem Statement:

You are given a sorted list of unique positive integers. Some numbers may be missing in the sequence. Your task is to find the smallest missing positive number using binary search, not linear search.

Example:

Input: [1, 2, 3, 5, 6, 7]
Output: 4

Input: [2, 3, 4, 7, 8]
Output: 1

Requirements:

- Use binary search to solve the problem efficiently.
- The solution should run in O(log n) time.
- Avoid using any loops that scan the array linearly.

In [11]:
def missing_seq(arr):

    left, right = 0, len(arr)-1

    while left <= right:
        mid = (left + right) //2
        if arr[mid] == mid + 1:
            left = mid+1
        else:
            right = mid-1
            
    return left+1

missing_seq([1,4,5,6,8])

2

**Longest Increasing Subarray** – contiguous elements that are strictly increasing

Example: [1, 2, 2, 3, 4] → [2, 3, 4]

In [12]:
def longest_seq_subarray(arr):
    start = 0
    max_len = 1
    max_start = 0
    for i in range(1, len(arr)):
        if arr[i] <= arr[i-1]:
            start = i

        curr_len = i - start + i
        if max_len < curr_len:
            max_len = curr_len
            max_start = start

    return arr[max_start : max_start + max_len]

longest_seq_subarray([1, 2, 2, 3, 4])

[2, 3, 4]

# Binary Search

1. You are given a sorted array nums of integers and a target value.
Write a function to return the index of the target if it exists in the array.
If it does not exist, return -1.

Input: nums = [1, 3, 5, 7, 9], target = 5  
Output: 2


In [13]:
def target_index(arr, target):
    left, right = 0, len(arr)-1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] > target:
            right = mid-1
        else:
            left = mid + 1
    return -1

target_index([1, 5, 6, 7, 9], 5)

1

2. Given a sorted array with duplicates, write a function to find the first occurrence of a target value.
Return the index if found, else -1.

Input: nums = [1, 2, 2, 2, 3], target = 2  
Output: 1


In [14]:
def first_index(arr, target):
    left, right = 0, len(arr)-1 
    idx = -1
    while left <= right:
        mid = (left + right) // 2 
        if arr[mid] == target:
            idx = mid
            right = mid - 1
        elif arr[mid] > target:
            right = mid - 1
        else :
            left = mid + 1
    return idx

first_index([1, 1, 2, 2, 3], 2)

2

3. Given a rotated sorted array (no duplicates), 
write a function to find the index of the minimum element.

Input: [4,5,6,7,0,1,2]  

Output: 4


Input: [2, 3, 4, 5, 6, 7, 1]  

Output: 6  # index of 1


In [15]:
def idx_of_min(arr):
    left, right = 0, len(arr) - 1
    while left < right:
        mid = (left + right) // 2
        if arr[mid] > arr[right]:
            left = mid + 1
        else:
            right = mid

    return left

print(idx_of_min([2, 3, 4, 5, 6, 7, 1, 0]))

7


4. Given a sorted array, write a function to find the smallest element greater than a given target.
Return its index; if no such element exists, return -1.

Input: nums = [1, 3, 5, 7], target = 4  

Output: 2  # nums[2] = 5


In [16]:
def small_el_than_target(nums, target):
    left, right = 0, len(nums) - 1
    ans = -1
    while left < right:
        mid = (left + right) // 2
        if nums[mid] > target:
            ans = mid
            right = mid - 1
        elif nums[mid] <= target:
            left = mid + 1
    return ans

small_el_than_target(nums = [1, 3, 5, 7], target = 4)

2

# Linked List

1. You are given a singly linked list.
Write a function to find the length of the linked list (i.e., count the number of nodes).

Input: 1 -> 2 -> 3 -> None  
Output: 3


In [17]:
def len_linked_list(ll):
    current = ll.head
    count = 0
    while current:
        current = current.next 
        count += 1
    return count

ll = LinkedList([1,2,3])
len_linked_list(ll)


3

2. Write a function to reverse a singly linked list.

Input: 1 -> 2 -> 3 -> None  
Output: 3 -> 2 -> 1 -> None


In [18]:
def reverse_ll(ll):
    current = ll.head
    prev = None
    while current:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node
    ll.head = prev

    return ll

ll = LinkedList([1,2,3])
reverse_ll(ll)

[3 -> 2 -> 1]

3. Write a function to detect a cycle in a singly linked list.
Return True if a cycle exists, else False.

    Input: 1 -> 2 -> 3 -> 4 -> 2 (cycle back to node 2)  
    Output: True


In [19]:
def if_cycle(ll):
    slow = fast = ll.head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True

    return False
        

4. Write a function to remove the nth node from the end of a singly linked list.

Input: 1 -> 2 -> 3 -> 4 -> 5, n = 2  
Output: 1 -> 2 -> 3 -> 5  # removed 4 (2nd from end)


In [20]:
def remove_nth_node_fromEnd(ll, n):
    dummy = Node(0)
    dummy.next = ll.head
    ll.head = dummy
    slow = fast = ll.head

    for _ in range(n+1):
        fast = fast.next

    while fast:
        slow = slow.next
        fast = fast.next

    slow.next = slow.next.next
    ll.head = dummy.next
    
    return ll.head

ll = LinkedList([1,2,3,4,5])
remove_nth_node_fromEnd(ll, 2)
ll

[1 -> 2 -> 3 -> 5]

# Array

1. You are given an array of integers.
Write a function to find the largest element in the array.

Input: [3, 7, 2, 9, 5]  
Output: 9


In [21]:
def largest(arr):
    if not arr:
        return False
    max_val =arr[0] 
    for i in arr:
        if i > max_val:
            max_val = i

    return max_val

largest([3, 7, 2, 9, 5])

9

2. You are given an array of integers.
Write a function to move all zeros to the end 
while maintaining the relative order of non-zero elements.

Input: [0, 1, 0, 3, 12]

Output: [1, 3, 12, 0, 0]


In [22]:
def move_zeroes(arr):
    pos = 0
    for i in range(len(arr)):
        if arr[i] != 0:
            arr[pos], arr[i] = arr[i], arr[pos]
            pos += 1

    return arr

move_zeroes([0, 1, 0, 3, 12])

[1, 3, 12, 0, 0]

3. Given an array of integers, write a function to find the subarray with the maximum sum (Kadane’s Algorithm).

Input: [-2, 1, -3, 4, -1, 2, 1, -5, 4]  
Output: 6  # max sum subarray: [4, -1, 2, 1]


In [23]:
def max_sum_subarray(arr):
    best_here = arr[0]
    best_all = float('-inf')
    for i in range(1, len(arr)):
        best_here = max(arr[i], best_here+arr[i])
        best_all = max(best_here, best_all)
    return best_all


def max_subarray(arr):
    best_here = arr[0]
    best_all = arr[0]

    start = 0
    best_start = 0
    best_end = 0
    
    for i in range(1, len(arr)):
        if arr[i] > best_here + arr[i]:
            best_here = arr[i]
            start = i
        else:
            best_here = best_here + arr[i]

        if best_here > best_all:
            best_all = best_here
            best_start = start
            best_end = i

    return arr[start : best_end + 1]


sum(max_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4])) == max_sum_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4])
        

True

4. Write a function to find all duplicates in an array of integers where:
- Numbers are in the range 1 to n
- Array size is n
- Do it in O(n) time and O(1) extra space (modify the array if needed)


Input: [4,3,2,7,8,2,3,1]  
Output: [2,3]


In [24]:
def duplicate_elements(arr):
    count = {} 
    for i in arr:
        count[i] = count.get(i, 0) + 1
    return [digit for digit, count in count.items() if count > 1]


def duplicate_elemnts_1_to_len(arr):
    dups = []

    for i in arr:
        idx = abs(i) - 1
        if arr[idx] < 0:
            dups.append(abs(i))
        else:
            arr[idx] = -arr[idx]
    return dups
    

duplicate_elements([4,3,2,7,8,2,3,1])
duplicate_elemnts_1_to_len([4,3,2,7,8,2,3,1])

[2, 3]

# String

1. Write a function to count the number of vowels (a, e, i, o, u) in a given string.

Input: "hello world"  
Output: 3


In [25]:
def count_vowels(s):
    vow = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
    count = 0
    for l in s:
        if l in vow:
            count += 1
    return count

count_vowels("hello world")

3

2. Write a function to check if two strings are anagrams of each other 
(contain the same characters in any order).

Input: s1 = "listen", s2 = "silent"  
Output: True


In [26]:
def is_anagram(s1, s2):
    if len(s1) != len(s2):
        return False

    from collections import Counter
    count = Counter(s1)
    for l in s2:
        if count[l] < 1:
            return False
        count[l] -= 1
    return all(value == 0 for value in count.values())

is_anagram("listen", "silent")

True

3. Write a function to find the first non-repeating character in a string.
Return the character itself; if none exists, return 'None'.

Input: "swiss"

Output: "w"


In [27]:
def first_non_duplicate(s):
    freq = {}
    for char in s:
        freq[char] = freq.get(char, 0) + 1

    for char in s:
        if freq[char] == 1:
            return char

    return None

first_non_duplicate("swiss")

'w'

4. Longest Substring Without Repeating Characters

👉 Problem:
Given a string s, find the length of the longest substring that has no repeating characters.

Input: s = "abcabcbb"

Output: 3

Explanation: "abc" is the longest substring without repeating characters.


In [28]:
def length_of_longest_substring(s):
    max_len = 1
    for i in range(len(s)):
        for j in range(1, len(s)):
            substr = s[i : j+1]

            if len(substr) == len(set(substr)):
                max_len = max(max_len, len(substr))

    return max_len

length_of_longest_substring("abcabcbb")

3

# Recursion

1. Write a recursive function to compute the factorial of a non-negative integer n.

Input: 5  
Output: 120


In [29]:
def factorial(num):
    if num < 0 or not isinstance(num, int):
        raise ValueError("factorial() expects only positive integers")
    
    if num <= 1:
        return 1

    return num * (factorial(num -1))

factorial(5)

120

2. Write a recursive function to find the sum of all elements in an array.

Input: [1, 2, 3, 4]

Output: 10


In [30]:
def sum_arr(arr, i=0):
    if i == len(arr):
        return 0
    return arr[i] + sum_arr(arr, i+1)

sum_arr([1,2,3,5])

11

3. Write a recursive function to find the nth Fibonacci number.

Input: n = 6  
Output: 8   # (0, 1, 1, 2, 3, 5, 8)

In [31]:
def nth_fibonacci(n):

    if n == 0: 
        return 0
    if n == 1:
        return 1

    return nth_fibonacci(n-1) + nth_fibonacci(n-2)

nth_fibonacci(6)

8

4. Implement Binary search using recursion

In [32]:
def binary_search_recursion(arr, target, left = 0, right = None ):
    if right is None:
        right = len(arr) - 1
        
    if left > right:
        return -1

    mid = (left + right) // 2
    if arr[mid] == target:
        return mid

    elif arr[mid] > target :
        binary_search_recursion(arr, target, left, mid - 1)

    else:
        binary_search_recursion(arr, target, mid + 1, right)


binary_search_recursion([1,2,3,4], 2)

1

In [33]:
sll = LinkedList([1,2,3,4,4,5,49])
dll = DoublyLinkedList([12,12,43,28,49,49])
# dll.insert(3,sll)
dll


def splice_SLL_in_DLL(sll, dll, index):
    if sll.head is None:
        print("empty Singly Linked List")
        return
    
    if dll.head is None:
        print("Empty Doubly Linked List")
        return

    current = dll.head
    count = 0
    while current.next and count < index:
        current = current.next
        count += 1

    temp = current.next
    current.next = sll.head
    sll.head.prev = current
    
    sll_curr = sll.head
    while sll_curr.next:
        sll_curr = sll_curr.next

    sll_curr.next = temp
    if temp:
        temp.prev = sll_curr


def remove_dups_inSLL(dll):

    cur = dll.head
    sll_head = None

    while cur and cur.next:
        if cur.next.prev is None:
            sll_head = cur
            print("sll found")
            break
        cur = cur.next

    if sll_head is None:
        return

    seen = set()
    s = sll_head
    prev = None
    while s and not s.next.prev:
        if s.data in seen:
            prev.next = s.next
            s = s.next
        else:
            seen.add(s.data)
            prev = s
            s = s.next



splice_SLL_in_DLL(sll,dll,3)
remove_dups_inSLL(dll)

dll

sll found


[12 <-> 12 <-> 43 <-> 28 <-> 1 <-> 2 <-> 3 <-> 4 <-> 5 <-> 49 <-> 49 <-> 49]